### DISCUSSION TITLE
Porting Moddable SDK to other Microcontrollers

### DISCUSSION THREAD

**Discussion by dhruvshah-volansys** [Moddable-OpenSource/moddable#616] [Porting]
**Description** 
I'm interested to know the process of adding support for a new target microcontroller as part of the Moddable SDK?

**Details Expected**
1. Step by Step Guide for adding a new target platform
2. Moddable SDK Software components diagram and their dependencies 
3. Moddable SDK Build process and any supporting documentation
4. Known constraint and/or facts which needs to be considered during porting
5. Guide to adding the support for the new interface (example: CAN)

**Open Queries**
1. Licensing details for Moddable SDK? Is it mandatory to commit the platform support back to this repo? 
2. What are the pre-requisites that chipset BSP shall suffice for integration of Moddable SDK?
3. Is it possible to limit the support for a target platform on a specific host system?

Let me know in case additional details are required.

**Comment by phoddie** [comment]
Your points cover a lot of ground. Here's some quick responses.

> 1. Step by Step Guide for adding a new target platform

This is impractical in general because each platform is unique  -- in  the capabilities available and in the build process.

The Moddable SDK is not a monolith. It is the XS JavaScript engine sitting on a very thin portable API layer. Just about all other functionality in the Moddable SDK is provided in the form of JavaScript modules. Each port is fundamentally about getting XS working on the target and then proceeding on to the modules that you want to support in that port.

The [XS Platforms](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/XS%20Platforms.md) document describes how the XS implementation is ported.

While not required, we bring up support for the xsbug debugger as early as possible in the porting process. Having the debugger is very helpful for understanding failures and the instrumentation view gives helpful insights into resource use. The debugging connection is typically serial, though it is not required to be. For example, it is possible to debug over a TCP socket connection.

As [noted](https://github.com/Moddable-OpenSource/moddable/issues/553#issuecomment-778533342) in an early discussion of the Pico port, it would be a good project for someone to put together porting notes based on the Pico effort.

> 2. Moddable SDK Software components diagram and their dependencies

XS itself is written in ANSI C and has relatively few dependencies beyond standard C library functions -- mostly memory and math. The JavaScript language itself, for example, does not have an file i/o built-in so stdio is not required.

For any given module, you can quickly determine its dependencies by looking at the `import` statements in the JavaScript source code to see what other modules it uses. You can also check module's manifest, for modules that provide one, to see what other modules it includes.

> 3. Moddable SDK Build process and any supporting documentation

Our documentation for building begins in our [Getting Started](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md) document. It describe how to build the SDK tools on macOS, Windows, and Linux. There is a separate document for each embedded target.

The heart of the build is `mcconfig`. It parses the project manifests and outputs make files. As you can imagine, the details differ significantly from one embedded target to the next. For some ports, it is necessary to modify `mcconfig` to support a particular features. The handling of `sdkconfig` on ESP32 targets is a good example of that. On the other hand, some ports require no changes to `mcconfig.` The Pico port is an example of that.

> 4. Known constraint and/or facts which needs to be considered during porting

The main considerations are the resource limitations of the target device:

- RAM
- flash storage
- CPU

An efficient port needs to manage these with care.

> 5. Guide to adding the support for the new interface (example: CAN)

Adding new hardware interface generally means adding a new module. For I/O, the Moddable SDK is moving to the [IO Class Pattern](https://ecmatc53.github.io/spec/web/spec.html#-9-io-class-pattern) from the forthcoming [ECMAScript® Embedded Systems API Specification](https://ecmatc53.github.io/spec/web/spec.html). Any new modules IO should follow this design pattern.

> 1.a Licensing details for Moddable SDK?

Please see our [licensing page](https://www.moddable.com/license).

> 1b. Is it mandatory to commit the platform support back to this repo?

It is not required to commit changes to this repository. Depending on the licensing option you choose, it may be required to provide the changes publicly in some fashion.

That said, we encourage those creating new ports to share them with the community that helped to build the Moddable SDK by making them available here.

> 2. What are the pre-requisites that chipset BSP shall suffice for integration of Moddable SDK?

That depends on what you want to do. We have [demonstrated](https://blog.moddable.com/blog/moddable-tech-presentation-to-tc-39/) the Moddable SDK running on devices with as little as 32 KB of RAM and 256 KB of flash.

> 3. Is it possible to limit the support for a target platform on a specific host system?

I'm not sure what you are asking.
